pkg/power/platform/source/redfish.go

```diff
@@ -0,0 +1,286 @@
+	PowerPowerSupplyReset PowerPowerSupplyReset `json:"#Power.PowerSupplyReset,omitempty"`
+}
+
+// RedfishSystemPowerResult is the system power query result
+type RedfishSystemPowerResult struct {
+	system        string
+	consumedWatts int
+	timestamp     time.Time
+}
+
+// RedfishAccessInfo is the struct for the access model
+type RedfishAccessInfo struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Host     string `json:"host"`
+}
+
+type RedFishClient struct {
+	// systemEnergy is the system accumulated energy consumption in Joule
+	accessInfo    RedfishAccessInfo
+	systems       []*RedfishSystemPowerResult
+	ticker        *time.Ticker
+	probeInterval time.Duration
+	mutex         sync.Mutex
+}
+
+func NewRedfishClient() *RedFishClient {
+	credPath := config.GetRedfishCredFilePath()
+	if credPath == "" {
+		klog.Infof("failed to get redfish credential file path")
+		return nil
+	}
+	if err := nodecred.InitNodeCredImpl(map[string]string{"redfish_cred_file_path": credPath}); err != nil {
+		klog.Infof("%s", fmt.Sprintf("failed to initialize node credential: %v", err))
+		return nil
+	} else {
+		klog.V(5).Infof("Initialized node credential")
+		nodeName := os.Getenv("NODE_NAME")
+		if nodeName == "" {
+			nodeName = "localhost"
+		}
+		redfishCred, err := nodecred.GetNodeCredByNodeName(nodeName, "redfish")
+		if err == nil {
+			userName := redfishCred["redfish_username"]
+			password := redfishCred["redfish_password"]
+			host := redfishCred["redfish_host"]
+			if userName != "" && password != "" && host != "" {
+				klog.V(5).Infof("Initialized redfish credential")
+				probeInterval := config.GetRedfishProbeIntervalInSeconds()
+				interval := time.Duration(probeInterval) * time.Second
+				redfish := &RedFishClient{
+					accessInfo:    RedfishAccessInfo{Username: userName, Password: password, Host: host},
+					systems:       []*RedfishSystemPowerResult{},
+					probeInterval: interval,
+					mutex:         sync.Mutex{},
+				}
+				return redfish
+			}
+		} else {
+			klog.V(1).Infof("%s", fmt.Sprintf("failed to get node credential: %v", err))
+			return nil
+		}
+	}
+	return nil
+}
+
+func (rf *RedFishClient) IsSystemCollectionSupported() bool {
+	system, err := getRedfishSystem(rf.accessInfo)
+
+	if err != nil {
+		klog.Infof("failed to get redfish system info: %v\n", err)
+		return false
+	}
+
+	intervalInMin := 0
+	// iterate each "Members" in the system and get the power info
+	for _, member := range system.Members {
+		// split the OdataID by delimiter "/" and get the system ID
+		split := strings.Split(member.OdataID, "/")
+		if len(split) < 2 {
+			continue
+		}
+		id := split[len(split)-1]
+		res := RedfishSystemPowerResult{}
+		power, err := getRedfishPower(rf.accessInfo, id)
+		if err == nil && len(power.PowerControl) > 0 {
+			res.system = id
+			res.consumedWatts = power.PowerControl[0].PowerConsumedWatts
+			res.timestamp = time.Now()
+			rf.systems = append(rf.systems, &res)
```

Comment:
I mean that it should replace `consumeWatts` only if the system had already existed in `rf.systems`.
```suggestion
	for index, member := range system.Members {
		// split the OdataID by delimiter "/" and get the system ID
		split := strings.Split(member.OdataID, "/")
		if len(split) < 2 {
			continue
		}
		id := split[len(split)-1]
		res := RedfishSystemPowerResult{}
		power, err := getRedfishPower(rf.accessInfo, id)
		if err == nil && len(power.PowerControl) > 0 {
			if index < len(rf.systems) {
				rf.systems[index].consumedWatts = power.PowerControl[0].PowerConsumedWatts
			} else {
				res.system = id
				res.consumedWatts = power.PowerControl[0].PowerConsumedWatts
				res.timestamp = time.Now()
				rf.systems = append(rf.systems, &res)
			}
```